NumLib/ODESolver/NonlinearSolver-impl.h

```diff
@@ -43,13 +43,15 @@ solve(Vector &x)
     auto& rhs   = MathLib::GlobalVectorProvider<Vector>::provider.getVector(_rhs_id);
     auto& x_new = MathLib::GlobalVectorProvider<Vector>::provider.getVector(_x_new_id);
 
-    bool success = false;
+    bool error_norms_met = false;
 
     BLAS::copy(x, x_new); // set initial guess, TODO save the copy
 
     for (unsigned iteration=1; iteration<_maxiter; ++iteration)
     {
-        sys.assembleMatricesPicard(x_new);
+        sys.preIteration(iteration, x);
+
+        sys.assembleMatricesPicard(x);
```

Comment:
At that point those two vectors have exactly the same values. But I think, it is more natural to call the assembly with something old (x) and not with something to be computed (x_new).
That's also more consistent with the Newton case.